### ISSUE TITLE
[WEB] AutoTabsRouter dynamic routes

### ISSUE THREAD

**Issue by xrax** [Milad-Akarie/auto_route_library#1197]
Hi! First of all, thanks so much for the auto route lib, it's amazing.

I'm using `AutoTabsRouter.tabBar` with dynamic routes that could be more or less depending on some gating logic. This is working fine just with a

```dart
AutoTabsRouter.tabBar(
  routes: gatingBloc().homeTabs, // <-- getter with dynamic routes logic
  builder: ...,
);
```

where the app can show 3 to 6 tabs depending on the gating logic.

I've noticed an issue where using the "full screen" button on a browser window triggers the flutter error red screen and I got in console:
![console](https://user-images.githubusercontent.com/12051271/185768329-d57b5320-ab87-47c7-9044-92fe6022b963.png)

Following the stack seems like what is happening is:

1. `AutoTabView` gets an updated widget ->
![autoTabView](https://user-images.githubusercontent.com/12051271/185768378-edf33213-7764-48b0-af6f-50850e906f63.png)

and triggers a `jumpToPage`.

2. `PageView.jumpToPage` tries to call `position.getPixelsFromPage`
![jumpToPage](https://user-images.githubusercontent.com/12051271/185768414-ed630292-bd57-416d-835d-b14f0f3927eb.png)

3. `position.getPixelsFromPage` tries to get the `viewPortDimension` that is null
![getPixelsFromPage](https://user-images.githubusercontent.com/12051271/185768444-36620d59-e0db-4384-bff2-ce64d456012d.png)

After many tries to get a workaround, I've found that if `gatingBloc().homeTabs` is returning always the same tabs the issue doesn't happens.

Does this sounds like a real issue? Or is it me trying to use the `AutoTabsRouter.tabBar` for something it is not intended? Any workaround? Any idea?

Here my flutter doctor -> 
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.5 21G72 darwin-x64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.2)
[✗] Cannot determine if IntelliJ is installed
    ✗ Directory listing failed
[✓] Connected device (2 available)
[✓] HTTP Host Availability
```

Thanks in advance.

**Comment by Milad-Akarie** [comment]
@xrax AutoTabsRouter should handle dynamic tabs, but not sure why is this happening on web, I'll check it out.

**Comment by xrax** [comment]
Hi @Milad-Akarie, I have an update to this error, `viewportDimension` on null at `didUpdateWidget` is not longer happening, it seems to be fixed with [5.0.1](https://pub.dev/packages/auto_route/versions/5.0.1) version. 😄 

There is a new error though, here the [error log](https://github.com/Milad-Akarie/auto_route_library/files/9640596/error-log-dynamic-routes.log) with the stack. Trigger error seems to be:

> This _AutoTabsRouterTabBar widget cannot be marked as needing to build because the framework is already in the process of building widgets. A widget can be marked as needing to be built during the build phase only if one of its ancestors is currently building. This exception is allowed because the framework builds parent widgets before children, which means a dirty descendant will always be built. Otherwise, the framework might not visit this widget during this build phase.

I've found a workaround just adding a `GlobalKey` to `AutoTabsRouter.tabBar` when dynamic routes has changed to force a rebuild having into account new params. Something like:

```
GlobalKey? _tabsRouterKey;
List<PageRouteInfo>? _routes;

void _checkRoutes(List<PageRouteInfo>? newRoutes) {
    if (newRoutes != null && !ListEquality().equals(_routes, newRoutes)) {
      _routes = newRoutes;
      _tabsRouterKey = GlobalKey();
    }
  }

...
  StreamBuilder<List<PageRouteInfo>>(
     stream: gatingBloc().homeTabsStream,
     initialData: gatingBloc().homeTabs,
     builder: (context, snapshot) {
       _checkRoutes(snapshot.data);

       return AutoTabsRouter.tabBar(
         key: _tabsRouterKey,
         routes: _routes!,
         builder: ...,
       );
     }
   );
```

It worked! The only downside with this approach is the `AutoTabsRouter` here loose its current `_activeIndex` and returns to `homeIndex`, I think make sense since I'm forcing it to be recreated. I could also kind of workaround it with a
```
AutoTabsRouter.of(context).setActiveIndex(
  savedTabIndex(),
);
```

Being called inside of the TabBar `builder` and `savedTabIndex` the new index for the old selected tab.

This approach is still less than ideal since the entire `TabBar` perform a "blink" from user perspective and also the final re-navigation is visible.

I have tried to find the cause in the `auto_route_library` code and tried many different approach like passing `tabController` in the `tabBar` constructor, to dispose `_tabController` at `_AutoTabsRouterTabBarState`  dispose or just at the beginning of `_updateTabController` and few other more, with no luck in any of those. I hope these new details helps to find any solution. 

Please let me know if I can help in any way to provide more details or to try other things.

Thanks a lot!
